File: source/uModbus.pas
unit uModbus;

{$IFDEF VER220}
  {$DEFINE UNDER_DELPHI_XE}
{$ENDIF}
{$IFDEF VER230}
  {$DEFINE UNDER_DELPHI_XE}
{$ENDIF}

interface
uses
 SysUtils, cport, WinSock, Sockets{$IFDEF UNDER_DELPHI_XE}, CPortTypes{$ENDIF};

type
  TModbusError = (
    merNone,
    merCantOpenPort,     // com port
    merTimeout,
    merAddressError,     // переданный и полученный адрес не совпадают
    merFuncError,        // переданный и полученный код функции не совпадают
    merCRCError,
    merWrongToLength,    // <253
    merWrongFromLength,  // from controller
    merModbusRecError,   // ошибка от контроллера, переданная им по модбасу
    merWrongDataLenFrom, // неправильная длина данных, полученная от контроллера
    merWrongDataLenTo,   // неправильная длина данных, переданная контроллеру
    merMBFunctionCode,   //-\
    merMBDataAddress,    //-|- ошибки, которые возвращает modbus
    merMBDataValue,      //-|
    merMBFuncExecute,    //-/
    merMFLenError,       // ошибки при приеме пакетов и преобразовании данных в модуле MiFARE
    merLast,
    merUndef = 255);

const
  StrModbusError: array [merNone .. merLast] of string = (
    'нет',
    'Не удалось открыть СОМ порт',
    'Таймаут связи с устройством',
    'Переданный и полученный адрес не совпадают',
    'Переданный и полученный код функции не совпадают',
    'Ошибка CRC пакета от устройства',
    'Слишком большой пакет для передачи устройству',
    'Несоответствие длины пакета и констант в пакете от устройства',
    'Ошибка модбаса',
    'Неправильная длина данных, полученная от устройства',
    'Неправильная длина данных, при попытке передачи на устройство',
    'Ошибка модбас (1) неправильная функция',
    'Ошибка модбас (2) неправильный адрес',
    'Ошибка модбас (3) неправильные данные в функции',
    'Ошибка модбас (4) исполнения функции',
    'Ошибка при приеме пакетов или преобразовании данных в модуле MiFARE',
    'нет'
    );

type
  TAddrQtyRec = packed record
   Addr: Word;
   Qty: Word;
  end;

  TAddrDataRec = packed record
   Addr: Word;
   Qty: Word;
  end;

  TRegDataRec = packed record
   Addr: Word;
   Qty: Word;
   data: array[1..250] of byte;
  end;

  TResultRec = packed record
   Qty: byte;
   data: array[1..250] of byte;
  end;

  TModbusIdentification = packed record
   Company,
   Product,
   Version,
   VendorURL: string;
  end;

  TModbus = class
  private
    FModbusTCP: boolean;
    FPort: TComPort;
    FSocket: TTcpClient;

    function SendTCP(TxSleep: cardinal; datain: AnsiString; var dataout: AnsiString): TModbusError;
    function SendRTU(TxSleep: cardinal; datain: AnsiString; var dataout: AnsiString): TModbusError;
  public
    constructor Create(Port: Integer; Speed: integer); overload;
    constructor Create(Host: string; Port: integer);   overload;
    destructor Destroy; reintroduce;

    function Send(ATxSleep: cardinal; Address, Func: byte; data: AnsiString; var dataout: AnsiString): TModbusError;

    function ReadCoilStatus(Address: byte; param: TAddrQtyRec; var output: TResultRec): TModbusError;      // 1
    function ReadDiscreteInputs(Address: byte; param: TAddrQtyRec; var output: TResultRec): TModbusError;  // 2
    function ReadHoldingRegisters(Address: byte; param: TAddrQtyRec; var output: TResultRec): TModbusError;// 3
    function ForceSingleCoil(Address: byte; OutputAddr: Word; state: boolean): TModbusError;               // 5
    function PresetSingleRegister(Address: byte; param: TAddrDataRec): TModbusError;                       // 6
    function WriteMultipleRegisters(Address: byte; input: TRegDataRec): TModbusError;                      //16

    //  modbus function
    function CallFunction(Address: byte; code, subcode: byte; data: AnsiString; var dataout: AnsiString): TModbusError;

    //  search device
    function FindFirst: byte;
    function FindNext(FromID: byte): byte;

    // Work with TCP
    function TCPConnect: boolean;

    function Connected: boolean;
  end;

implementation

uses Math, windows;

{ TModbus }

function ArrByteToString(data: array of byte; Qty: integer = -1): ansistring;
var
 len: integer;
begin
  Result := '';
  if Qty < 0 then
    len := length(data)
  else
    len := Qty;

  if len > 250 then exit;

  SetLength(Result, len);
  move(data[0], Result[1], len);
end;

function TModbus.CallFunction(Address, code, subcode: byte;
  data: AnsiString; var dataout: AnsiString): TModbusError;
begin
  Result := Send(100, Address, code, ansichar(subcode) + data, dataout);
end;

constructor TModbus.Create(Port, Speed: integer);
begin
  inherited Create;

  FModbusTCP := false;
  FPort := TComPort.Create(Nil);
  FPort.Port := 'COM' + IntToStr(Port);

  case speed of
    115200: FPort.BaudRate := br115200;
    9600: FPort.BaudRate := br9600;
    2400: FPort.BaudRate := br2400;
    1200: FPort.BaudRate := br1200;
  else
    FPort.BaudRate := br115200;
  end;
  try
    FPort.Open;
  except
  end;
end;

function TModbus.Connected: boolean;
begin
  if not FModbusTCP then
    Result := FPort.Connected
  else
    Result := TCPConnect;
end;

constructor TModbus.Create(Host: string; Port: integer);
begin
  inherited Create;

  FModbusTCP := true;
  FSocket := TTcpClient.Create(nil);
  FSocket.RemoteHost := Host;
  FSocket.RemotePort := IntToStr(Port);
  FSocket.BlockMode := bmBlocking;
end;

destructor TModbus.Destroy;
begin
  if FModbusTCP then
  begin
    FSocket.Close;
    FSocket.Destroy;
  end
  else
  begin
    FPort.Close;
    FPort.Destroy;
  end;

  inherited;
end;

function crc16string(msg: ansistring): ansistring;
var
  crc: word;
  n,
  i: integer;
  b:byte;
begin
  crc := $FFFF;
  for i := 1 to length (msg) do
  begin
    b := byte(msg[i]);
    crc := crc xor b;
    for n := 1 to 8 do
    begin
      if (crc and 1) <> 0 then
        crc := (crc shr 1) xor $A001
      else crc := crc shr 1;
    end;
  end;

  result := ansichar(crc and $ff) + ansichar(crc shr 8);
end;

function TModbus.FindFirst: byte;
var
  i: integer;
  err: TModbusError;
  param: TAddrQtyRec;
  output: TResultRec;
begin
  Result := 0;
  param.Addr := 0;
  param.Qty := 1;
  for i := 1 to 250 do
  begin
    err := ReadHoldingRegisters(i, param, output);
    if not (err in [merCantOpenPort, merTimeout, merWrongDataLenTo]) then
    begin
      Result := i;
     exit;
    end;
  end;
end;

function TModbus.FindNext(FromID: byte): byte;
var
  i: integer;
  err: TModbusError;
  param: TAddrQtyRec;
  output: TResultRec;
begin
  Result := 0;
  param.Addr := 0;
  param.Qty := 1;
  for i := FromID + 1 to 250 do
  begin
    err := ReadHoldingRegisters(i, param, output);
    if not (err in [merCantOpenPort, merTimeout, merWrongDataLenTo]) then
    begin
     Result := i;
     exit;
    end;
  end;
end;

function TModbus.ForceSingleCoil(Address: byte; OutputAddr: Word; state: boolean): TModbusError;
begin
  Result := merNone
end;

function TModbus.PresetSingleRegister(Address: byte; param: TAddrDataRec): TModbusError;
begin
  Result := merNone
end;

function TModbus.ReadCoilStatus(Address: byte; param: TAddrQtyRec;
  var output: TResultRec): TModbusError;
var
 vin,
 vout: AnsiString;
begin
  output.Qty := 0;
  vin := '    ';
  move(param, vin[1], 4);
  Result := Send(20 + param.Qty div 12, Address, 1, vin, vout);

  if Result = merNone then
  begin
   if (length(vout) < 2) or (ord(vout[1]) + 1 <> length(vout))  then
   begin
    Result := merWrongDataLenFrom;
    exit;
   end;

   output.Qty := Ord(vout[1]);
   move(vout[2], output.data[1], output.Qty);
  end;
end;

function TModbus.ReadDiscreteInputs(Address: byte; param: TAddrQtyRec;
  var output: TResultRec): TModbusError;
begin
  Result := merNone
end;

function TModbus.ReadHoldingRegisters(Address: byte; param: TAddrQtyRec;
  var output: TResultRec): TModbusError;
var
  vin,
  vout: AnsiString;
begin
  Result := merWrongDataLenTo;
  if param.Qty > 128 then exit;
  vin := AnsiChar(Hi(param.Addr)) + AnsiChar(Lo(param.Addr)) +
    AnsiChar(Hi(param.Qty)) + AnsiChar(Lo(param.Qty));

  Result := Send(40 + param.Qty div 2, Address, 03, vin, vout);
  if Result <> merNone then exit;
  if length(vout) <> ord(vout[3]) + 5 then
  begin
    Result := merWrongDataLenFrom;
    exit;
  end;

  output.Qty := ord(vout[3]);
// SetLength(output.data, output.Qty);
  move(vout[4], output.data[1], output.Qty);
end;

function TModbus.Send(ATxSleep: cardinal; Address, Func: byte; data: AnsiString; var dataout: AnsiString): TModbusError;
var
  s: ansistring;
begin
  Result := merUndef;
  dataout := '';

  if Length(data) >= 253 then
  begin
    Result := merWrongToLength;
    exit;
  end;

  s := AnsiChar(Address) + AnsiChar(Func) + data;

  // send
  try
    if FModbusTCP then
      Result := SendTCP(ATxSleep, s, dataout)
    else
      Result := SendRTU(ATxSleep, s, dataout);

    if Result <> merNone then exit;
  except
    sleep(10);
  end;

  // process
  if Length(dataout) = 0 then
  begin
    Result := merTimeout;
    exit;
  end;

  if (Length(dataout) < 4) or (Length(dataout) > 250) then
  begin
    dataout := '';
    Result := merWrongFromLength;
    exit;
  end;

  if dataout[1] <> AnsiChar(Address) then
  begin
    Result := merAddressError;
    exit;
  end;
  if dataout[2] <> AnsiChar(Func) then
  begin
    Result := merFuncError;
    if (ord(dataout[2]) xor $80) = Func then
    begin
      case ord(dataout[3]) of
        01: Result := merMBFunctionCode;
        02: Result := merMBDataAddress;
        03: Result := merMBDataValue;
        04: Result := merMBFuncExecute;
        else
      end;
    end;
    exit;
  end;
end;

function TModbus.SendRTU(TxSleep: cardinal; datain: AnsiString; var dataout: AnsiString): TModbusError;
var
  s: AnsiString;
  cnt: integer;
  i, j: integer;
  tmp: cardinal;
begin
  Result := merNone;
  s := datain + crc16string(datain);

  if not FPort.Connected then
  begin
    Result := merCantOpenPort;
    exit;
  end;

  //flush
  FPort.ReadStr(dataout, FPort.InputCount);
  //write

  FPort.WriteStr(s);

  TxSleep := GetTickCount + TxSleep;
  while (GetTickCount < TxSleep) do
  begin
    cnt := FPort.InputCount;
    if cnt <> 0 then
    begin
      tmp := GetTickCount;
      if TxSleep > tmp then
        sleep(TxSleep - tmp);
      break;
    end;
  end;

//  sleep(TxSleep);
  dataout := '';
//  cnt := FPort.InputCount;
  if cnt <> 0 then
    FPort.ReadStr(dataout, cnt)
  else
    Result := merTimeout;

  // похоже выгребли не все
  i := 5;
  while ((dataout = '') or (crc16string(dataout) <> #0#0)) and (i > 0) do
  begin
    sleep(20);

    FPort.ReadStr(s, FPort.InputCount);
    dataout := dataout + s;
    if s <> '' then
      Result := merNone;

    i := i - 1;
  end;

  if Result <> merNone then exit;
  if crc16string(dataout) <> #0#0 then
  begin
    Result := merCRCError;
    exit;
  end;
end;

function TModbus.SendTCP(TxSleep: cardinal; datain: AnsiString; var dataout: AnsiString): TModbusError;
var
  cnt: integer;
  s: AnsiString;
  buf: array [0 .. 1024] of AnsiChar;
  len: word;
begin
  Result := merNone;
  len := length(datain);
  s := AnsiString(#0#1) + AnsiString(#0#0) + AnsiChar(hi(len)) + AnsiChar(lo(len)) + datain;

  if not FSocket.Connected then
   try
     FSocket.Close;
     if not TCPConnect then
      begin
       Result := merCantOpenPort;
       exit;
      end;
   except
    Result := merCantOpenPort;
    exit;
   end;

  //flush
  if FSocket.WaitForData(1) then
   begin
    cnt := FSocket.PeekBuf(buf[0], length(buf));
    if cnt > 0 then FSocket.ReceiveBuf(buf[0], cnt);
   end;

  //write
  cnt := FSocket.SendBuf(s[1], length(s));

  // try to resend if socket closed
  if cnt = -1 then
   begin
    FSocket.Close;
    if TCPConnect then
      cnt := FSocket.SendBuf(s[1], length(s));
   end;

  if cnt <> length(s) then
   begin
    Result := merCantOpenPort;
    FSocket.Close;
    exit;
   end;

  dataout := '';

  if FSocket.WaitForData(TxSleep) then
   begin
    cnt := FSocket.PeekBuf(buf[0], length(buf));
    if cnt > 0 then
     begin
      FSocket.ReceiveBuf(buf[0], cnt);
      SetLength(dataout, cnt);
      move(buf[0], dataout[1], cnt);
     end
    else
      Result := merTimeout;
   end;

  // похоже выгребли не все
  if dataout = '' then
   if FSocket.WaitForData(200) then
    begin
     cnt := FSocket.PeekBuf(buf[0], length(buf));
     if cnt > 0 then
      begin
       FSocket.ReceiveBuf(buf[0], cnt);
       SetLength(s, cnt);
       move(buf[0], s[1], cnt);
      end;

     if s <> '' then Result := merNone;
     dataout := dataout + s;
    end;

  if length(dataout) = 0 then
   begin
    Result := merTimeout;
    exit;
   end;

  if length(dataout) < 9 then
   begin
    dataout := '';
    Result := merWrongFromLength;
    exit;
   end;

  if ord(dataout[5]) * 256 + ord(dataout[6]) <> length(dataout) - 6 then
   begin
    dataout := '';
    Result := merWrongFromLength;
    exit;
   end;

  dataout := Copy(dataout, 7, length(dataout));
  dataout := dataout + crc16String(dataout);
end;

function TModbus.TCPConnect: boolean;
begin
  Result := false;
  if not FModbusTCP then
  begin
    Result := true;
    exit;
  end;
  try
    Result := FSocket.Connect
  except
  end;
end;

function TModbus.WriteMultipleRegisters(Address: byte;
  input: TRegDataRec): TModbusError;
var
  vin: AnsiString;
  output: ansistring;
begin
  Result := merWrongDataLenTo;
  if input.Qty > 128 then exit;
  vin := AnsiChar(Hi(input.Addr)) + AnsiChar(Lo(input.Addr)) +
    AnsiChar(Hi(input.Qty)) + AnsiChar(Lo(input.Qty)) +
    AnsiChar(input.Qty * 2) +
    ArrByteToString(input.data, input.Qty * 2);

  Result := Send(40 + input.Qty, Address, 16, vin, output);

  if Result <> merNone then exit;
  if ord(output[5]) * $FF + ord(output[6]) <> input.Qty then
  begin
    Result := merWrongDataLenFrom;
    exit;
  end;
end;

end.

File: TurniketTest_C/uModbus.cpp
#include "uTypes.h"
#include "uModbus.h"
#include "uSerial.h"
#include "uTtcp.h"
#include "uCrc16.h"

const char * StrModbusError[merLast+1] =
{
    "OK",
    "Can't open channel",
    "Device connection timeout",
    "Sent and received addresses are not equal",
    "Sent and received func codes are not equal",
    "Device packet CRC error",
    "Too long packet for device",
    "Packet length error",
    "Modbus error",
    "Wrong data length from device",
    "Wrong data length for device",
    "Modbus error (1) Illegal function",
    "Modbus error (2) Illegal data address",
    "Modbus error (3) Illegal data value",
    "Modbus error (4) Function execution",
    "MiFARE module error",
    "NONE"
};

TModbusError TModbus :: CallFunction(byte Address, byte code, byte subcode, bytesArray * data, bytesArray * dataout)
{
    bytesArray bdata;

    bdata.length = data->length +1;
    bdata.data[0] = subcode;
    for(int i=0; i< data->length; i++) bdata.data[i+1] = data->data[i];
    TModbusError err = Send(100, Address, code, &bdata, dataout);
    return err;
}


TModbus :: TModbus(int Port, int Speed)
{
    if(Port < 0 || Port > MAX_VALID_PORT)
    {
        FPort=NULL;
        ModbusError = merCantOpenPort;
    }
    else
    {
        FPort = new TComPort();
        if(FPort==NULL)
        {
            ModbusError = merCantOpenPort;
        }
        else
        {
            FModbusTCP = 0;
            ModbusError = merNone;
#ifdef QNX_X86
            sprintf(FPort->Port,DEVSERPATH"%d",Port);
#else
			sprintf(FPort->Port,DEVSERPATH"%d",Port);  // should be corrected for real system: Port-1 or Port
#endif            
            FTimeout = DEFAULT_TIMEOUT;
            switch(Speed)
            {
            case 115200:
                FPort->setBaudRate(B115200);
                break;
            case 9600:
                FPort->setBaudRate(B9600);
                break;
            case 2400:
                FPort->setBaudRate(B2400);
                break;
            case 1200:
                FPort->setBaudRate(B1200);
                break;
            default:
                FPort->setBaudRate(B115200);
                break;
            }
            if(FPort->Open()<0)
            {
                delete FPort;
                FPort = NULL;
                ModbusError = merCantOpenPort;
            }
        }
    }
}

bool TModbus :: Connect()
{
    if(!FModbusTCP) return RTUConnect();
    else return TCPConnect();
    return 0;
}


TModbus :: TModbus(char * Host, int Port, bool blockMode)
{
    if(Port < 1 || Port > MAX_TCP_PORT)
    {
        ModbusError = merCantOpenPort;
        FSocket = NULL;
    }
    else
    {
        FSocket = new TTcpClient(blockMode);
        if(FSocket==NULL)
        {
            ModbusError = merCantOpenPort;
        }
        else
        {
            ModbusError = merNone;
            FModbusTCP = 1;
            FTimeout = DEFAULT_TIMEOUT;
            strcpy(FSocket->RemoteHost,Host);
            sprintf(FSocket->RemotePort,"%d",Port);
            FSocket->setSockIP();
            FSocket->setSockPort();
            if(!(FSocket->getSockIP()) || !(FSocket->getSockPort()))
            {
                ModbusError = merCantOpenPort;
                delete FSocket;
                FSocket = NULL;
            }
            FSocket->setBlockMode(1);  // default mode for tcp communication
        }
    }
}


TModbus :: ~TModbus()
{
    if(FModbusTCP)
    {
        if(FSocket != NULL)
        {
            FSocket->Close();
            delete FSocket;
            FSocket=NULL;
        }
    }
    else
    {
        if(FPort!= NULL)
        {
            FPort->Close();
            delete FPort;
            FPort=NULL;
        }
    }
}


byte TModbus :: FindFirst()
{
    byte i;
    TAddrQtyRec param;
    TResultRec output;

    param.Addr = 0;
    param.Qty = 0;
    for(i=1;i<=250;i++)  // Valid modbus device addresses are from 1 to 247 !!!
    {
        ModbusError = ReadHoldingRegisters(i, &param, &output);
        if(ModbusError != merCantOpenPort &&
           ModbusError != merTimeout &&
           ModbusError != merWrongDataLenTo)
        {
            return i;
        }
        usleep(GetTimeOut()*1000);
    }
    return 0;
}

byte TModbus :: FindNext(byte FromID)
{
    byte i;
    TAddrQtyRec param;
    TResultRec output;
    param.Addr = 0;
    param.Qty = 1;
    for(i=FromID+1; i<=250; i++)  // Valid addresses are from 1 to 247 !!!
    {
        ModbusError = ReadHoldingRegisters(i, &param, &output);
        if(ModbusError!=merCantOpenPort &&
           ModbusError!= merTimeout &&
           ModbusError != merWrongDataLenTo)
        {
            return i;
        }
    }
    return 0;
}


/*

TModbusError TModbus :: ForceSingleCoil(byte Address, Word OutputAddr, boolean state)
{
    // TODO: should be implemented
    // Is not used in this application
    ModbusError = merNone;
    return merNone;
}

*/

/*

TModbusError TModbus :: PresetSingleRegister(byte Address, TAddrDataRec * param)
{
    // TODO: should be implemented
    // Is not used in this application
    ModbusError = merNone;
    return merNone;
}
*/

TModbusError TModbus :: ReadCoilStatus(byte Address, TAddrQtyRec * param, TResultRec * output)
{
    // Implemented but not used in this application

    bytesArray vin;
    bytesArray vout;

    output->Qty = 0;
    vin.length = 4;
    vin.data[0] = (byte)(((param->Addr) & 0xff00) >> 8);
    vin.data[1] = (byte)((param->Addr) & 0x00ff);
    vin.data[2] = (byte)(((param->Qty) & 0xff00) >> 8);
    vin.data[3] = (byte)((param->Qty) & 0x00ff);
    ModbusError = Send(20 + (param->Qty)/12, Address, 1, &vin, &vout);
    if(ModbusError==merNone)
    {
        if((vout.length < 6) || ((vout.data[2] + 5) != vout.length))
        {
            ModbusError = merWrongDataLenFrom;
            return merWrongDataLenFrom;
        }
        output->Qty = vout.data[3];  // Response's count
        for(byte i = 1; i <= output->Qty; i++) output->data[i-1] = vout.data[i+2]; // Response's data
    }
    return ModbusError;
}


/*
TModbusError TModbus :: ReadDiscreteInputs(byte Address, TAddrQtyRec * param, TResultRec * output)
{
    // TODO: should be implemented
    // Is not used in this application
    ModbusError = merNone;
    return merNone;
}
*/


TModbusError TModbus :: ReadHoldingRegisters(byte Address, TAddrQtyRec * param,  TResultRec * output)
{
    bytesArray vin;
    bytesArray vout;

    ModbusError = merWrongDataLenTo;
    if(param->Qty >= 128) return ModbusError;

    output->Qty = 0;
    vin.length = 4;
    vin.data[0] = (byte)(((param->Addr) & 0xff00) >> 8);
    vin.data[1] = (byte)((param->Addr) & 0x00ff);
    vin.data[2] = (byte)(((param->Qty) & 0xff00) >> 8);
    vin.data[3] = (byte)((param->Qty) & 0x00ff);

    ModbusError = Send(40 + (param->Qty)/2, Address, 3, &vin, &vout);
    if(ModbusError != merNone) return ModbusError;
    if(vout.length != (vout.data[2] + 5))
        {
            ModbusError = merWrongDataLenFrom;
            return ModbusError;
        }
    output->Qty = vout.data[2];
    for(byte i = 1; i <= output->Qty; i++) output->data[i-1] = vout.data[i+2]; // Response's data
    return ModbusError;
}




TModbusError TModbus :: Send(cardinal ATxSleep, byte Address, byte Func, bytesArray * data, bytesArray * dataout)
{
    bytesArray s;

    ModbusError = merNone;  // Clear initial error status
    dataout->length = 0;

    if(data->length >= 253)
    {
        ModbusError = merWrongToLength;
        return ModbusError;
    }

    s.data[0] = Address;
    s.data[1] = Func;
    for(int i=0; i<data->length; i++) s.data[i+2] = data->data[i];
	s.length = data->length+2;
    // Send
    if(FModbusTCP)
    {
    	LongWord tcp_atxsleep = (ATxSleep > MIN_TCP_ATXSLEEP) ? ATxSleep : MIN_TCP_ATXSLEEP;
        ModbusError = SendTCP(tcp_atxsleep,&s,dataout);
    }
    else
    {
        ModbusError = SendRTU(ATxSleep,&s,dataout);
    }
    if(ModbusError != merNone) return ModbusError;

      // Process
  if(!dataout->length)
  {
      ModbusError = merTimeout;
      return ModbusError;
  }
  if(dataout->length < 4 || dataout->length > 250)
  {
      dataout->length = 0;
      ModbusError = merWrongFromLength;
      return ModbusError;
  }
  if(dataout->data[0] != Address)
  {
      ModbusError = merAddressError;
      return ModbusError;
  }

  if(dataout->data[1] != Func)
  {
      if(((dataout->data[1]) ^ 0x80) == Func)
      {
          switch(dataout->data[2])
          {
              case 1:
                ModbusError = merMBFunctionCode;    // ILLEGAL FUNCTION
                break;
              case 2:
                ModbusError = merMBDataAddress;     // ILLEGAL DATA ADDRESS
                break;
              case 3:
                ModbusError = merMBDataValue;       // ILLEGAL DATA VALUE
                break;
              case 4:
                ModbusError = merMBFuncExecute;     // Error while function execution
                break;
              default:
                ModbusError = merTimeout;
                break;
          }
      }
      else ModbusError = merFuncError; // Invalid function code in response packet
  }
  else ModbusError = merNone;

  return ModbusError;
}

TModbusError TModbus :: SendRTU(cardinal TxSleep, bytesArray * datain, bytesArray * dataout)
{
  bytesArray s;
  bytesArray * request = &s;
  bytesArray * response = dataout;
  int i;

  md_crc wCRC16, tCRC16;

  if(FPort==NULL)
  {
     return (ModbusError =  merCantOpenPort);
  }
  if(!FPort->getConnected())
  {
     return (ModbusError =  merCantOpenPort);
  }

  ModbusError = merNone;
  s.length = datain->length + 2;
  wCRC16.d = tCRC16.d = 0;
  wCRC16 = modbus_crc16(&tCRC16, datain);
  for(i=0; i<datain->length; i++) 
  {
  	s.data[i] = datain->data[i]; 
  }
  for(i=0; i<2; i++) 
  {
  	s.data[datain->length+i] = wCRC16.b[i]; 
  }

 //flush
  FPort->Flush();
 //send request
  FPort->writeToSerPort(request);
 //wait for response
  FPort->waitForAnswer(TxSleep);
 //get response
  if(FPort->getInputCount() > 0)
  {
    FPort->readFromSerPort(response);
  }
  else
  {
    return (ModbusError = merTimeout);
  }
  //check crc
  wCRC16.d = tCRC16.d = 0;
  wCRC16 = modbus_crc16(&tCRC16, response);
  if(wCRC16.d != 0)
  {
      return (ModbusError = merCRCError);
  }
  else
  {
      return (ModbusError = merNone);
  }

  return ModbusError;
}

TModbusError TModbus :: SendTCP(cardinal TxSleep, bytesArray * datain, bytesArray * dataout)
{
    int cnt;
    bytesArray s;
    md_crc wCRC16, tCRC16;
    bytes2word len;
	
	dataout->length = 0;
    ModbusError = merNone;
    len.d = datain->length;
    s.length = datain->length + 6;
    s.data[0] = 0;  // Transaction ID (Hi)
    s.data[1] = 1;  // Transaction ID (Lo)
    s.data[2] = s.data[3] = 0;  // Modbus TCP protocol ID == 0
#if  defined(QNX_X86) || (defined(__BYTE_ORDER) && __BYTE_ORDER == __LITTLE_ENDIAN) || !defined(__BYTE_ORDER)
    s.data[4] = len.b[1];  // Hi
    s.data[5] = len.b[0];  // Lo
#else  // big-endian otherwise
    s.data[4] = len.b[0];  // Hi
    s.data[5] = len.b[1];  // Lo
#endif
    for(int i = 0; i < datain->length; i++) 
    {
    	s.data[i+6] = datain->data[i];
    }
    if(!FSocket->getConnected())
    {
        FSocket->Close();
        if(!TCPConnect())
        {
           return (ModbusError = merCantOpenPort);
        }
    }
    FSocket->Flush();
    // send request  
    cnt = FSocket->writeToTcpSocket(&s);
    // try to resend if socket closed
     if(cnt<=0)
     {
      FSocket->Close();
      if(TCPConnect())
      {
          cnt = FSocket->writeToTcpSocket(&s);
      }
     }
    
    if(cnt != (int)(s.length))
    {
      FSocket->Close();
      return (ModbusError = merCantOpenPort);
    }

    // wait and read answer
    if( FSocket->waitAndReadAnswer(TxSleep) <= 0)
    {
        //  0   incompleet packet
        //  -1  read Socket error
        //  -2  ModbusTCP header error (may be dust)
        return (ModbusError = merTimeout);
    }
    FSocket->readFromBuff(dataout);
  	if(dataout->length < 9)
  	{
      dataout->length = 0;
      return (ModbusError = merWrongFromLength);
  	}
    if(((Word)(dataout->data[4]))*256 + (Word)(dataout->data[5]) != (dataout->length - 6))
	  {
	      dataout->length = 0;
	      return (ModbusError = merWrongFromLength);
	  }
    for(int i = 6; i < dataout->length; i++)
	  {
	    dataout->data[i-6] = dataout->data[i];
	  }
	dataout->length = dataout->length - 6;
	  // Add crc16
	wCRC16.d = tCRC16.d = 0;
	wCRC16 = modbus_crc16(&tCRC16, dataout);
	dataout->data[dataout->length] = wCRC16.b[0];
	dataout->data[dataout->length+1] = wCRC16.b[1];
	dataout->length = dataout->length + 2;

	return (ModbusError = merNone);
}


bool TModbus :: TCPConnect()
{
    bool rc = 0;
    if(!FModbusTCP)
    {
        return rc;
    }
    rc = FSocket->Connect();
    return rc;
}

bool TModbus :: RTUConnect()
{
    bool rc = 0;
    if(FModbusTCP)
    {
        return rc;
    }
    FPort->Open();
    return FPort->getConnected();
}


TModbusError TModbus :: WriteMultipleRegisters(byte Address, TRegDataRec * input)
{
    bytesArray vin;
    bytesArray output;

    ModbusError = merWrongDataLenTo;

    if(input->Qty >= 128) return ModbusError;

    vin.length = 5 + input->Qty * 2;
    vin.data[0] = (byte)(((input->Addr) & 0xff00) >> 8);
    vin.data[1] = (byte)((input->Addr) & 0x00ff);
    vin.data[2] = (byte)(((input->Qty) & 0xff00) >> 8);
    vin.data[3] = (byte)((input->Qty) & 0x00ff);
    vin.data[4] = (byte)((input->Qty)*2);
    for(int i = 0; i < (input->Qty)*2; i++) vin.data[i+5] = input->data[i];  // like ArrByteToString

    ModbusError = Send(10 + FTimeout + input->Qty, Address, 16, &vin, &output);

    if(ModbusError != merNone) return ModbusError;
    if((((Word)(output.data[4]))*256 + (Word)(output.data[5])) != input->Qty)
        {
            ModbusError = merWrongDataLenFrom;
        }
    return ModbusError;
}

